### ISSUE TITLE
VoxCPM2 Chirp/Click Artifact & Voice Consistency in One-Shot Cloning

### ISSUE THREAD

**Issue by dcox761** [OpenBMB/VoxCPM#272]
Thank you for all your hard work on VoxCPM and the recent VoxCPM2 release. So far I am very impressed with the results for a project I am working on. Using with nanovllm-voxcpm on an AWS EC2 g6.xlarge (NVidia L4 24GB) I am achieving streaming responses within 250ms and LoRA training completes 2000 steps in approx. 15min.

Are you able to provide some guidance for a small but annoying issue?

One-shot voice cloning on the base VoxCPM2 model produces a chirp/click at the start of every generated audio segment - this appears to be the tail of the reference audio leaking through the DiT's `prefix_feat_cond` conditioning. For example if my reference audio ends with "enemy" the generated audio appears to start with "emy". 

While I have attempted changes through nanovllm-voxcpm as described below, I can reproduce this issue easily with the default VoxCPM LoRA WebUI.

I do have a mix of audio segments that appear to work fine with these changes for some voices but am unable to work out the main factor could be causing this.

**Things attempted:**
1. **Blank audio padding on reference clip** - Did not help; last audio patch still used as DiT conditioning, silence padding made it worse
2. **Reference mode** (send `ref_audio_latents_base64` only) - Eliminated chirp but significantly degraded voice quality; reverted
3. **Zero `prefix_feat_cond` patch** (to nanovllm-voxcpm) - Zeroed the DiT conditioning during prefill to match training behaviour; did not completely resolve
4. **Dual mode - ref_audio + prompt latents together** - Send both for maximum voice conditioning; did not completely resolve
5. **PCM chirp trim - 100ms** (HACK) - Pipeline-level trim of first ~2 patches when cloning is active; addresses residual LM continuation artifact but still does not work completely
6. **Transcript mismatch** - Obviously extra words in the reference text cause problems but even when reviewed carefully the Chirp/Click exists

**Comment by codeHackeR321** [comment]
I am using HiFi-Cloning and for me sometimes the pace / speed of output audio is too fast, like it feels speaker speaks too fast. Otherwise it is normal / relevant to the input voice. 
@dcox761 Have you also experiences something similar ?

**Comment by madey83** [comment]
> I am using HiFi-Cloning and for me sometimes the pace / speed of output audio is too fast, like it feels speaker speaks too fast. Otherwise it is normal / relevant to the input voice. [@dcox761](https://github.com/dcox761) Have you also experiences something similar ?

How is long text used for generation? When I generate audio for subtitles, I haven't noticed any issues with speed. I've noticed that when the audio is clean, it doesn't produce any chirps or clicks at the beginning. However, when I use a file that isn't so clean, it introduces problems at the start of almost every sentence. This is related to version 2.0.2.

**Comment by akshat-schmooze** [comment]
> > I am using HiFi-Cloning and for me sometimes the pace / speed of output audio is too fast, like it feels speaker speaks too fast. Otherwise it is normal / relevant to the input voice. [@dcox761](https://github.com/dcox761) Have you also experiences something similar ?
> 
> How is long text used for generation? When I generate audio for subtitles, I haven't noticed any issues with speed. I've noticed that when the audio is clean, it doesn't produce any chirps or clicks at the beginning. However, when I use a file that isn't so clean, it introduces problems at the start of almost every sentence. This is related to version 2.0.2.

I generate 1 sentence per request, hence it is 2-10 words per request, not much longer. I am also using 2.0.2.

**Comment by madey83** [comment]
My venv has PyTorch 2.7.x with CUDA 12.8, and using this setup, I’ve created 3-4 voice-overs of 650 lines each.

**Comment by codeHackeR321** [comment]
> My venv has PyTorch 2.7.x with CUDA 12.8, and using this setup, I’ve created 3-4 voice-overs of 650 lines each.

I am using cuda 12.4 and torch 2.5.1. I'll upgrade and check.

**Comment by dcox761** [comment]
I have had some audio clips work using zero prefix_feat_cond patch to nanovllm-voxcpm and ref_audio + prompt latents together with carefully checked transcript when encoding.

Audio clips from 2 to 20sec were used for LoRA training but I found the problem reproduced with one-shot cloning against the base VoxCPM2 model.

Generating TTS for single sentences from just a few words in size. They either all chirp or are all good.

Some reference audio of 6 to 20sec is good while others are not. I can't work out what it is about the "bad" audio that causes the chirp.
The recordings appear to be clear and consistent with no noise.

VoxCPM2 has been a big improvement for me as I used to have issues with changing speed and volume.

I am building VoxCPM into the nanovllm-voxcpm Docker image, currently VoxCPM 2.0.0 using PyTorch 2.10.0 and CUDA 12.8.

For now I have stuck with the hack, PCM chirp trim - 100ms.

**Comment by madey83** [comment]
I do not have any problem with additional "noise" on generated audio, but still fights with cutted off ends of the words. I can't find any pattern.

**Comment by yoimiya122** [comment]
I found that when using the 2.0.2 codebase (pulled from the branch and built/installed locally), and ensuring the reference audio has 0.5-1.0 seconds of silence at the end, the issue where the tail of the reference audio gets appended to the generated audio no longer occurs when using Ultimate Clone. Maybe you can try it.

**Comment by bluecrystaladmin** [comment]
> I found that when using the 2.0.2 codebase (pulled from the branch and built/installed locally), and ensuring the reference audio has 0.5-1.0 seconds of silence at the end, the issue where the tail of the reference audio gets appended to the generated audio no longer occurs when using Ultimate Clone. Maybe you can try it.

2.0.2 improved this for me and having tail padding appeared to help.
